src/file.rs

```diff
@@ -1,117 +1,556 @@
+    if natts == 0 {
+        return Ok(HashMap::new());
+    }
+    let mut attributes = HashMap::with_capacity(natts as _);
+    for i in 0..natts {
+        let mut buf = [0u8; NC_MAX_NAME as usize + 1];
+        let err = unsafe { nc_inq_attname(ncid, varid, i, buf.as_mut_ptr() as *mut _) };
+
+        if err != NC_NOERR {
+            return Err(err.into());
+        }
+        let name: Vec<_> = buf
+            .iter()
+            .take_while(|x| **x != 0)
+            .map(|x| *x as u8)
+            .collect();
+        let name = String::from_utf8(name).unwrap();
+        let a = Attribute {
+            name: name.clone(),
+            ncid,
+            varid,
         };
-    init_group(&mut root);
-    Ok(File {
-        id: ncid, 
-        name: file.to_string(),
-        root: root,
-    })
+        attributes.insert(name, a);
+    }
+
+    Ok(attributes)
 }
 
-/// Open a netCDF file in creation mode (write only).
-pub fn create(file: &str) -> Result<File, String> {
-    let data_path = path::Path::new(file);
-    let f = ffi::CString::new(data_path.to_str().unwrap()).unwrap();
-    let mut ncid : i32 = -999999i32;
-    let err : i32;
+fn get_dimensions_of_var(
+    ncid: nc_type,
+    varid: nc_type,
+    unlimited_dims: &[nc_type],
+) -> error::Result<Vec<Dimension>> {
+    let mut ndims = 0;
+    let err;
     unsafe {
-        let _g = libnetcdf_lock.lock().unwrap();
-        err = nc_create(f.as_ptr(), NC_NETCDF4, &mut ncid);
+        err = nc_inq_var(
+            ncid,
+            varid,
+            std::ptr::null_mut(),
+            std::ptr::null_mut(),
+            &mut ndims,
+            std::ptr::null_mut(),
+            std::ptr::null_mut(),
+        );
     }
     if err != NC_NOERR {
-        return Err(NC_ERRORS.get(&err).unwrap().clone());
+        return Err(err.into());
     }
-    let root = Group {
-            name: "root".to_string(),
-            id: ncid,
-            variables: HashMap::new(),
-            attributes: HashMap::new(),
-            dimensions: HashMap::new(),
-            sub_groups: HashMap::new(),
+    if ndims == 0 {
+        return Ok(Vec::new());
+    }
+    let mut dimids = vec![0; ndims as usize];
+    let err;
+    unsafe {
+        err = nc_inq_var(
+            ncid,
+            varid,
+            std::ptr::null_mut(),
+            std::ptr::null_mut(),
+            std::ptr::null_mut(),
+            dimids.as_mut_ptr(),
+            std::ptr::null_mut(),
+        );
+    }
+    if err != NC_NOERR {
+        return Err(err.into());
+    }
+
+    let mut dimensions = Vec::with_capacity(ndims as usize);
+    for dimid in dimids.into_iter() {
+        let mut name = vec![0u8; NC_MAX_NAME as usize + 1];
+        let mut dimlen = 0;
+        let err;
+        unsafe {
+            err = nc_inq_dim(ncid, dimid, name.as_mut_ptr() as *mut _, &mut dimlen);
+        }
+
+        if err != NC_NOERR {
+            return Err(err.into());
+        }
```

Comment:
It seems like this little snippet would be good to factor out into a function. So that you can write `check_nc_error(err)?;` instead, except with a better function name. `check_nc_error` would be a `-> error::Result<()>` function.

Reply:
I named it checked, so all uses will be
```rust
error::checked(nc_do_something())?
```